src/BaseMailer.php

```diff
@@ -0,0 +1,284 @@
+     */
+    private $logger;
+
+    /**
+     * @param MessageFactoryInterface $messageFactory
+     * @param Composer $composer
+     * @param EventDispatcherInterface $eventDispatcher
+     * @param LoggerInterface $logger
+     */
+    public function __construct(MessageFactoryInterface $messageFactory, Composer $composer, EventDispatcherInterface $eventDispatcher, LoggerInterface $logger)
+    {
+        $this->messageFactory = $messageFactory;
+        $this->composer = $composer;
+        $this->eventDispatcher = $eventDispatcher;
+        $this->logger = $logger;
+    }
+
+    /**
+     * @return Composer message composer instance.
+     */
+    public function getComposer(): Composer
+    {
+        return $this->composer;
+    }
+
+    /**
+     * Creates a new message instance and optionally composes its body content via view rendering.
+     *
+     * @param string|array|null $view the view to be used for rendering the message body. This can be:
+     *
+     * - a string, which represents the view name for rendering the HTML body of the email.
+     *   In this case, the text body will be generated by applying `strip_tags()` to the HTML body.
+     * - an array with 'html' and/or 'text' elements. The 'html' element refers to the view name
+     *   for rendering the HTML body, while 'text' element is for rendering the text body. For example,
+     *   `['html' => 'contact-html', 'text' => 'contact-text']`.
+     * - null, meaning the message instance will be returned without body content.
+     *
+     * The view to be rendered can be specified in one of the following formats:
+     *
+     * - a relative view name (e.g. "contact") located under [[viewPath]].
+     *
+     * @param array $params the parameters (name-value pairs) that will be extracted and made available in the view file.
+     * @return MessageInterface message instance.
+     */
+    public function compose($view = null, array $params = []): MessageInterface
+    {
+        $message = $this->createMessage();
+        if ($view === null) {
+            return $message;
+        }
+
+        $this->composer->compose($message, $view, $params);
+
+        return $message;
+    }
+
+    /**
+     * Creates a new message instance.
+     * @return MessageInterface message instance.
+     */
+    protected function createMessage(): MessageInterface
+    {
+        return $this->messageFactory->create($this);
+    }
+
+    /**
+     * Sends the given email message.
+     * This method will log a message about the email being sent.
+     * If [[useFileTransport]] is true, it will save the email as a file under [[fileTransportPath]].
+     * Otherwise, it will call [[sendMessage()]] to send the email to its recipient(s).
+     * Child classes should implement [[sendMessage()]] with the actual email sending logic.
+     * @param MessageInterface $message email message instance to be sent
+     * @return bool whether the message has been sent successfully
```

Comment:
Potentially there could be multiple reasons for failure. How about switching to exceptions instead of returning boolean?

Reply:
We can also defines another event(`error`) to achieve this purpose, pseudocode on `BaseMailer::send`:

```php
try {
    ...
    $isSuccessful = $this->sendMessage($message);
    ...
} catch {\Throwable $e} {
    $this->onError($message, $e);
    return false;
}
```

One benefit of this is that it doesn't need developer to try/catch, how do you think?

Reply:
>  it doesn't need developer to try/catch

But you have try-catch in your example above...

Reply:
Umm, I mean we don't need to try/catch on every time we send mail:

```php
try {
    $mailer->composer()->...->send();
} catch (ExceptionOne $e) {
   ...
} catch (ExceptionTwo $e) {
   ...
} catch (\Throwable $e) {
   ...
}
```

Instead, implement `try/catch` in `BaseMailer::send` and let event handler to handle exceptions.

Reply:
In the event handler you have to do it anyway. Either with multiple `catch`, or multiple `else` / `case` in the handler.

Reply:
OK, it seems this way does not make much sence. I'll try to finish it ASAP.